tools/lib/src/device.dart

```diff
@@ -0,0 +1,369 @@
+    }
+    return device;
+  }
+
+  /// Creates a new reference to a Tizen emulator device.
+  ///
+  /// Valid characters for a device name are [A-Za-z0-9-_].
+  factory Device.emulator(
+    String name,
+    Profile profile, {
+    required TizenSdk tizenSdk,
+    SyncableProcessRunner processRunner = const SyncableProcessRunner(),
+  }) =>
+      EmulatorDevice(
+        name,
+        profile,
+        tizenSdk: tizenSdk,
+        processRunner: processRunner,
+      );
+
+  /// The name of the device.
+  final String name;
+
+  /// A profile of the device.
+  final Profile profile;
+
+  final TizenSdk _tizenSdk;
+
+  String? _id;
+
+  final SyncableProcessRunner _processRunner;
+
+  /// The unqiue identifier assigned to a connected device.
+  String? get id => _id;
+
+  /// Whether this device is an emulator.
+  bool get isEmulator => false;
+
+  /// Whether this device is connected to host PC.
+  bool get isConnected {
+    _id = _findId();
+    return _id != null;
+  }
+
+  static final RegExp _logPattern =
+      RegExp(r'\d\d:\d\d\s+([(\+\d+\s+)|(~\d+\s+)|(\-\d+\s+)]+):\s+(.*)');
+
+  String? _findId() {
+    final List<SdbDeviceInfo> deviceInfos = _tizenSdk.sdbDevices();
+    for (final SdbDeviceInfo deviceInfo in deviceInfos) {
+      if (deviceInfo.name == name) {
+        return deviceInfo.id;
+      }
+    }
+    return null;
+  }
+
+  /// Runs integration test in [workingDir].
+  ///
+  /// [workingDir] must be a valid exisiting flutter directory where
+  /// `flutter pub get` has already been ran succesfully. For an app project,
+  /// [workingDir] is the app's source root. For a plugin project, [workingDir]
+  /// is one of the example directories.
+  ///
+  /// If test doesn't finish after [timeout], it's considered a failure and the
+  /// function will return [PackageResult.fail] with time expired log.
+  ///
+  /// Returns [PackageResult.success] when every test passes succesfully,
+  /// otherwise returns [PackageResult.fail]. Never returns [PackageResult.skip]
+  /// nor [PackageResult.exclude].
+  Future<PackageResult> runIntegrationTest(
+    Directory workingDir,
+    Duration timeout,
+  ) async {
+    if (!isConnected) {
+      return PackageResult.fail(
+          <String>['Device $name($profile) is not connected.']);
+    }
+
+    final io.Process process = await _processRunner.start(
+      'flutter-tizen',
+      <String>['-d', id!, 'test', 'integration_test'],
+      workingDirectory: workingDir,
+    );
+
+    bool timedOut = false;
+    final Stream<String> streamLines = process.stdout
+        .transform(const Utf8Decoder())
+        .transform(const LineSplitter())
+        .timeout(
+      timeout,
+      onTimeout: (EventSink<String> sink) {
+        timedOut = true;
+        sink.close();
+      },
+    );
+
+    String lastLine = '';
+    final DateTime start = DateTime.now();
+    await for (final String line in streamLines) {
+      lastLine = line;
+      io.stdout.add('$line\n'.codeUnits);
```

Comment:
In this context, all of them do the same thing; print the test progress in the console. I think it's needed since each integration test takes some time, but I'll remove it if you feel the same way after running the tool locally without it.

Reply:
Oh, sorry. What I meant was the conversion to code units is redundant but not the line itself is. Either `print` or `stdout.writeln` (without `.codeUnits`) will be more readable.